File: Divisionals/K/submissions/accepted/krazy-darcy.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<pair<int,int> > > adj;
vector<bool> vis;
vector<string> ans;

void dfs(int v,int step, const string& my_answer){
  if(vis[v]) return;
  vis[v] = true;
  
  ans[v] = my_answer;

  for(auto u : adj[v])
    if(u.second <= step) // Am I allowed to use this edge?
      dfs(u.first,step,my_answer);
}

int main(){
  int n, x, y;
  cin >> n >> x >> y;

  adj.resize(n);
  for(int i=0;i<x;i++){
    int u, v;
    cin >> u >> v;
    u--;
    v--;
    adj[u].emplace_back(v,0);
    adj[v].emplace_back(u,0);
  }
  for(int i=0;i<y;i++){
    int u, v;
    cin >> u >> v;
    u--;
    v--;
    adj[u].emplace_back(v,0);
    adj[v].emplace_back(u,1);
  }

  ans.assign(n,"Impossible");
  
  vis.assign(n,false);
  dfs(0,1,"Ticket");
  
  vis.assign(n,false);
  dfs(0,0,"No Ticket");

  for(int v=0;v<n;v++)
    cout << ans[v] << endl;
}

File: Divisionals/K/submissions/accepted/KrazyDarcy.java
import java.util.*;

public class KrazyDarcy {
  public static void main(String[] args) {
    new KrazyDarcy().run();
  }

  int n;
  int[] D;
  int[][] A;

  void run(){
    Scanner in = new Scanner(System.in);
    
    n = in.nextInt();
    int x = in.nextInt(), y = in.nextInt();

    A = new int[n][n];

    for(int i=0;i<x;i++){
      int u = in.nextInt()-1, v = in.nextInt()-1;
      A[u][v] = A[v][u] = 2;
    }
    for(int i=0;i<y;i++){
      int u = in.nextInt()-1, v = in.nextInt()-1;
      A[u][v] = 2;
      if(A[v][u] == 0) A[v][u] = 1;
    }

    D = new int[n];

    dfs(0,1);
    dfs(0,2);

    String[] output = {"Impossible", "Ticket", "No Ticket"};

    for(int i=0;i<n;i++) System.out.println(output[D[i]]);
  }

  void dfs(int v, int val){
    D[v] = val;
    for(int i=0;i<n;i++)
      if(A[v][i] >= val && D[i] != val)
        dfs(i, val);
  }
}
